message/messaging_service.cc

```diff
@@ -643,13 +648,24 @@ shared_ptr<messaging_service::rpc_protocol_client_wrapper> messaging_service::ge
         auto& snitch_ptr = locator::i_endpoint_snitch::get_local_snitch_ptr();
 
         // either rack/dc need to be in same dc to use non-tls
-        if (snitch_ptr->get_datacenter(id.addr) != snitch_ptr->get_datacenter(utils::fb_utilities::get_broadcast_address())) {
+        auto my_dc = snitch_ptr->get_datacenter(broadcast_address);
+        if (snitch_ptr->get_datacenter(addr) != my_dc) {
+            return true;
+        }
+        // #9653 - if our idea of dc for bind address differs from our official endpoint address,
+        // we cannot trust downgrading.
+        if (broadcast_address != laddr && snitch_ptr->get_datacenter(laddr) != my_dc) {
```

Comment:
Nobody will understand the trick behind the check `snitch_ptr->get_datacenter(laddr) != my_dc` which is trying to probe if the get_datacenter understand the private address. Let's add more comments. 

Aslo, something like this will make it easier to understand. 

```
bool can_query_dc_using_private address = snitch_ptr->get_datacenter(laddr) == my_dc;
```

Reply:
I felt the comment said roughly the same, but I can either augment it or use a var named as suggested.

Reply:
For me, I do not really understand any of the comments until I realize the code is utilizing a bug in get_datacenter() to do the decision not to downgrade to non encryption.

Reply:
I am not sure I would call it "utilizing" a bug. It ensures that if the call does not work for addresses we try to use, we don't fail. It is a double check.